src/os/bluestore/BlueStore.cc

```diff
@@ -8093,30 +8094,58 @@ BlueStore::OnodeRef BlueStore::fsck_check_objects_shallow(
   o.reset(Onode::decode(c, oid, key, value));
   ++num_objects;
 
-  num_spanning_blobs += o->extent_map.spanning_blob_map.size();
-
-  o->extent_map.fault_range(db, 0, OBJECT_MAX_SIZE);
-  _dump_onode<30>(cct, *o);
-  // shards
   if (!o->extent_map.shards.empty()) {
     ++num_sharded_objects;
-    if (depth != FSCK_SHALLOW) {
-      ceph_assert(expecting_shards);
-      for (auto& s : o->extent_map.shards) {
-        dout(20) << __func__ << "    shard " << *s.shard_info << dendl;
-        expecting_shards->push_back(string());
-        get_extent_shard_key(o->key, s.shard_info->offset,
-          &expecting_shards->back());
-        if (s.shard_info->offset >= o->onode.size) {
-          derr << "fsck error: " << oid << " shard 0x" << std::hex
-            << s.shard_info->offset << " past EOF at 0x" << o->onode.size
-            << std::dec << dendl;
-          ++errors;
-        }
+    int64_t shard_errors = 0;
+    // we should have all the shards we expect to have, in order
+    uint32_t shard_id = 0;
+    for (auto& s : o->extent_map.shards) {
+      dout(20) << __func__ << "    shard " << *s.shard_info << dendl;
+      if (s.shard_info->offset >= o->onode.size) {
+	derr << "fsck error: " << oid << " shard 0x" << std::hex
+	     << s.shard_info->offset << " past EOF at 0x" << o->onode.size
+	     << std::dec << dendl;
+	++shard_errors;
+      }
+      string shard_key;
+      get_extent_shard_key(o->key, s.shard_info->offset, &shard_key);
+      if (auto it = shards_local.find(shard_key); it != shards_local.end()) {
+	// if we have such shard, decode
+	o->extent_map.provide_shard_info_to_onode(it->second, shard_id);
+	shard_id++;
+	shards_local.erase(it);
+      } else {
+	// if not, its missing
+	derr << "fsck error: " << oid << " shard 0x" << std::hex
+	     << s.shard_info->offset << std::dec << " missing" << dendl;
+	++shard_errors;
       }
     }
+    // any shards still in shard_kvals are extra
+    for (auto& s: shards_local) {
+      derr << "fsck error: " << oid << " stray shard "
+	   << pretty_binary_string(s.first) << dendl;
+      ++shard_errors;
+    }
+
+    errors += shard_errors;
+    if (shard_errors > 0) {
```

Comment:
This covers both lack of shard and stray shard cases. Do we really want to delete an object in case of stray shard(s)?

Reply:
The same applies to one more error case: s.shard_info->offset >= o->onode.size

Reply:
And I realize that this brings us back to the discussion about which is a better approach to recover from broken shard information error(s). Generally I can accept the "remove everything" approach when some shards are missed. But I still don't like the idea to remove onode when it looks rather valid and some additional/unexpected shards are present only. IMO still no need to be that destructive during repair in this case. The error(s) might be in e.g. shard removal procedure which just leaves some records intact. While onode consistency is perfectly fine...